### PR TITLE
db setup for unit testing

### PR DESCRIPTION
**Pull request by talhaw**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
 	${DOCKRUN} bash ./scripts/check.sh
 
-todo: prepare
+todo:
 	${DOCKRUN} bash ./scripts/todo.sh
 
-test: prepare format check
-	${DOCKRUN} bash ./scripts/test.sh
+db_seed:
+	docker run --name mysql-db -e MYSQL_ALLOW_EMPTY_PASSWORD=yes -p 3306:3306 -d mysql:5.6
```
Comment (ahab94) [inline Makefile]: This 
```
docker run --name mysql-db -e MYSQL_ALLOW_EMPTY_PASSWORD=yes -p 3306:3306 -d mysql:5.6
docker run --name mongo-db -p 27015:27015 -d mongo:4.2.0
```
should be `db_start`

Comment (talhaw) [reply Makefile]: okk

File: Makefile [hunk 2]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
+	docker run --name mongo-db -p 27015:27015 -d mongo:4.2.0
+	docker cp apna_school.sql mysql-db:apna_school.sql
```
Comment (ahab94) [inline Makefile]: And this
```
	docker cp apna_school.sql mysql-db:apna_school.sql
	docker exec mysql-db sh -c 'mysql -uroot < apna_school.sql'
```
should be `db_seed`
No need to seed mongo. Because mongo creates keyspaces if it doesn't exist automatically.

File: Makefile [hunk 3]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
+	docker exec mysql-db sh -c 'mysql -uroot < apna_school.sql'
+	docker exec mongo-db sh -c 'mongo < apna_school.mongo'
+
+db_stop:
+	docker stop mysql-db
+	docker stop mongo-db
+	docker rm -f mysql-db
+	docker rm -f mongo-db
+
+test: check db_seed db_start db_stop
```
Comment (ahab94) [inline Makefile]: The error you are facing could be because `db_stop` is a pre step to `db_test` Which is incorrect. db stop is post stop. After the tests have been concluded.

Comment (ahab94) [inline Makefile]: Also let's not crowd the targets with pre steps.
```
test: check db_seed db_start db_stop
```
Lets break it down like...
```
test: check db_start
```
and 
```
db_start: db_seed
```

`db_stop` should come after. See how to achieve running `db_stop` after `test`.

File: apna_school.mongo [hunk 4]
```diff
@@ -0,0 +1,7 @@
+use apna_school;
```
Comment (ahab94) [inline apna_school.mongo]: We don't need this file.

Comment (ahab94) [inline apna_school.mongo]: Lets delete this file.

File: Makefile [hunk 5]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
 
-test: prepare format check
+test: check db_seed
 	${DOCKRUN} bash ./scripts/test.sh
+
+db_start: db_stop
+	@docker run --name unit-testing-mysql-db -e MYSQL_ALLOW_EMPTY_PASSWORD=yes -p 3306:3306 -d mysql:5.6
+	@docker run --name unit-testing-mongo-db -p 27015-27017:27015-27017 -d mongo:4.2.0
+
+db_seed: db_start
```
Comment (ahab94) [inline Makefile]: We don't need mongo in db_seed

File: Makefile [hunk 6]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
-todo: prepare
+todo:
 	${DOCKRUN} bash ./scripts/todo.sh
 
-test: prepare format check
+test: check db_seed
 	${DOCKRUN} bash ./scripts/test.sh
+
+db_start: db_stop
+	@docker run --name unit-testing-mysql-db -e MYSQL_ALLOW_EMPTY_PASSWORD=yes -p 3306:3306 -d mysql:5.6
```
Comment (ahab94) [inline Makefile]: How about we name these containers:
`apna-school-mysql-db`
`apna-school-mongo-db`

File: Makefile [hunk 7]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
+	@docker run --name unit-testing-mongo-db -p 27015-27017:27015-27017 -d mongo:4.2.0
+
+db_seed: db_start
+	@docker cp apna_school.sql unit-testing-mysql-db:apna_school.sql
+	@docker cp apna_school.mongo unit-testing-mongo-db:apna_school.mongo
+	@echo "Executing databases...wait for 30 seconds"
+	@sleep 30
```
Comment (ahab94) [inline Makefile]: Can we verify that if we reduce tis to 20 or 15... does it work in that case?

Comment (ahab94) [reply Makefile]: Because this adds up 30 seconds to the build time... we want to keep it as minimal as we can.

File: Makefile [hunk 8]
```diff
@@ -19,11 +19,27 @@ tmp/dev_image_id:
...
+	@docker exec -i unit-testing-mysql-db sh -c 'mysql -uroot < apna_school.sql'
+	@docker exec -i unit-testing-mongo-db sh -c 'mongo < apna_school.mongo'
+
+db_stop:
```
Comment (ahab94) [inline Makefile]: There is one problem running db stop first:
```
Error: No such container: unit-testing-mysql-db
Makefile:44: recipe for target 'db_stop' failed
make: *** [db_stop] Error 1
```

Comment (ahab94) [reply Makefile]: It looks like we should verify if the container exists  or not first.

File: db/mysql/client.go [hunk 9]
```diff
@@ -24,7 +24,7 @@ func init() {
 	db.Register("mysql", NewClient)
 }
 
-//The first implementation.
+//The first implementation./
```
Comment (ahab94) [inline db/mysql/client.go]: Please revert this change.

File: db/mongo/client_test.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+	c, _ := NewClient(db.Option{})
+	tempObj := models.Teacher{Name: "Qamar Zaman Bajwa", Age: "12", Department: "Army", Salary: 430}
+	_, _ = c.AddTeacher(&tempObj)
+	type args struct {
+		id string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
```
Comment (ahab94) [inline db/mongo/client_test.go]: we need `want` in `Get` calls.

File: db/mongo/client_test.go [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+			}
+		})
+	}
+}
+
+func Test_client_GetTeacher(t *testing.T) {
+	os.Setenv("DB_PORT", "27017")
+	os.Setenv("DB_HOST", "apna-school-mongo-db")
+	c, _ := NewClient(db.Option{})
+	tempObj := models.Teacher{Name: "Qamar Zaman Bajwa", Age: "12", Department: "Army", Salary: 430}
```
Comment (ahab94) [inline db/mongo/client_test.go]: lets not call it `tempObj`
